### PR TITLE
extension_header

### LINKED ISSUE #3277: eBPF store APIs are not backward compatible.
#3110 modified the `ebpf_helper_function_prototype_t` struct and thus broke backwards compatiblity with extensions that were invoke the the eBPF store APIs to export program information data. #3263 temporary reverts the offending change. A backward compatible store API is required.

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

This fixes #3277. The extension data structures now have version and size fields that can be used to check compatibility with extensions. This will be a breaking change for all extensions. The eBPF API for exporting program data has been renamed as well.

## Testing

CI tasks. Later #2747 will be added as well.

## Documentation

`eBPFExtension.md` is modified.

## Installation

N/A.

### CODE REVIEW

**Review by Alan-Jowett — APPROVED** [review]

File: docs/eBpfExtensions.md [hunk 1]
```diff
@@ -62,8 +62,14 @@ initialized as follows:
...
   * The `data` field of this structure should point to a structure of type `ebpf_program_data_t`.
 
+#### `ebpf_extension_header_t` Struct
+This is a mandatory header that is common to all data structures needed by eBPF extensions to register with the eBPF framework.
+* `version`: Version of the extension data structure.
+* `size`: Size of the extension data structure.
+
 #### `ebpf_program_data_t` Struct
 The various fields of this structure should be set as follows:
+* `header`: Version () and size.
```
Comment (dthaler) [inline docs/eBpfExtensions.md:72]:
```suggestion
* `header`: Version and size.
```

File: docs/eBpfExtensions.md [hunk 2]
```diff
@@ -189,7 +199,14 @@ This optional function is used to populate the flat buffers representing the dat
 returned to the application when the `bpf_prog_test_run_opts` call completes. In addition, the function frees any
 resources allocated in the `ebpf_program_context_create_t` call.
 
-### 2.2 Program Information NPI Client Attach and Detach Callbacks
+### 2.2 Backward compatibility of the Extension data structures
+All the extension data structures are versioned. New fields can be added to the end of a data structure to maintain backward compatibility
+with the existing extensions. In such cases, the size field of the header will be updated but the version of the structure *will not change*.
+If the change in data structure is such that it is no longer backward compatible (such as change field type or position),
```
Comment (dthaler) [inline docs/eBpfExtensions.md:205]:
```suggestion
If the change in data structure is such that it is no longer backward compatible (such as changing field type or position),
```

File: docs/eBpfExtensions.md [hunk 3]
```diff
@@ -189,7 +199,14 @@ This optional function is used to populate the flat buffers representing the dat
...
+then the version number will be updated. All previous versions of extension structure will continue to be supported.
```
Comment (dthaler) [inline docs/eBpfExtensions.md:206]:
```suggestion
then the version number will be updated. All previous versions of the extension structure will continue to be supported.
```

File: include/ebpf_program_types.h [hunk 4]
```diff
@@ -4,22 +4,15 @@
...
-#endif
+#include "ebpf_windows.h"
 
 #define EBPF_MAX_PROGRAM_DESCRIPTOR_NAME_LENGTH 256
 #define EBPF_MAX_HELPER_FUNCTION_NAME_LENGTH 256
 
 typedef struct _ebpf_program_type_descriptor
 {
+    ebpf_extension_header_t header;
+    // The following fields are available in version EBPF_PROGRAM_TYPE_DESCRIPTOR_VERSION_0 and later.
```
Comment (dthaler) [inline include/ebpf_program_types.h:15]: This comment seems incorrect.  "and later" shouldn't apply to versions.  Same in line 26.  I'd replace these comments with comments saying this is the typedef when version = 0.

File: include/ebpf_windows.h [hunk 5]
```diff
@@ -49,3 +57,30 @@ typedef enum _ebpf_helper_function
...
+#define EBPF_PROGRAM_INFORMATION_VERSION_0 0
+#define EBPF_HELPER_FUNCTION_ADDRESSES_VERSION_0 0
+#define EBPF_PROGRAM_SECTION_INFORMATION_VERSION_0 0
+#define EBPF_PROGRAM_DATA_VERSION_0 0
+#define EBPF_PROGRAM_SECTION_VERSION_0 0
+
+/**
+ * @brief Header of an eBPF extension data structure.
+ * Every eBPF extension data structure must start with this header.
+ * New fields can be added to the end eBPF extension data structure
```
Comment (dthaler) [inline include/ebpf_windows.h:78]:
```suggestion
 * New fields can be added to the end of an eBPF extension data structure
```

File: libs/execution_context/ebpf_native.c [hunk 6]
```diff
@@ -121,7 +121,7 @@ static const NPI_PROVIDER_CHARACTERISTICS _ebpf_native_provider_characteristics
     _ebpf_native_provider_detach_client_callback,
     NULL,
     {
-        EBPF_PROGRAM_INFORMATION_PROVIDER_DATA_VERSION,
+        0,
```
Comment (dthaler) [inline libs/execution_context/ebpf_native.c:124]: Ok but in other places you use a *_VERSION_0 define.  For consistency, why not here?

File: libs/shared/shared_common.c [hunk 7]
```diff
@@ -1,8 +1,72 @@
...
 
+#include "ebpf_program_types.h"
 #include "ebpf_shared_framework.h"
 
+static bool
+_ebpf_validate_helper_function_prototype(const ebpf_helper_function_prototype_t* helper_prototype)
+{
+    return (
+        (helper_prototype != NULL) && (helper_prototype->header.size >= EBPF_HELPER_FUNCTION_PROTOTYPE_MINIMUM_SIZE) &&
+        (helper_prototype->header.version <= EBPF_HELPER_FUNCTION_PROTOTYPE_VERSION_LATEST) &&
```
Comment (dthaler) [inline libs/shared/shared_common.c:12]: BUG: This check is not correct.  The minimum size is per version, so since this is doing a size comparison against a constant, you can only do an exact comparison against version == VERSION_0, not "<=".   E.g., consider the case where the version increases and the size decreases.  The structure would be valid but this will incorrectly return invalid.

File: libs/shared/shared_common.c [hunk 8]
```diff
@@ -1,8 +1,72 @@
...
+            }
+        }
+    }
+    return true;
+}
+
+static bool
+_ebpf_validate_context_descriptor(_In_ const ebpf_context_descriptor_t* context_descriptor)
+{
+    return ((context_descriptor != NULL) && (context_descriptor->size >= sizeof(ebpf_context_descriptor_t)));
```
Comment (dthaler) [inline libs/shared/shared_common.c:33]: Why no version check here?

File: libs/shared/shared_common.c [hunk 9]
```diff
@@ -1,8 +1,72 @@
...
+}
+
+static bool
+_ebpf_validate_program_type_descriptor(_In_ const ebpf_program_type_descriptor_t* program_type_descriptor)
+{
+    return (
+        (program_type_descriptor != NULL) &&
+        (program_type_descriptor->header.size >= EBPF_PROGRAM_TYPE_DESCRIPTOR_MINIMUM_SIZE) &&
+        (program_type_descriptor->header.version <= EBPF_PROGRAM_TYPE_DESCRIPTOR_VERSION_LATEST) &&
```
Comment (dthaler) [inline libs/shared/shared_common.c:42]: BUG (see above)

File: libs/shared/shared_common.c [hunk 10]
```diff
@@ -1,8 +1,72 @@
...
+        (program_type_descriptor->name != NULL) &&
+        _ebpf_validate_context_descriptor(program_type_descriptor->context_descriptor));
+}
+
+bool
+ebpf_validate_program_info(_In_ const ebpf_program_info_t* program_info)
+{
+    return (
+        (program_info != NULL) && (program_info->header.size >= EBPF_PROGRAM_INFO_MINIMUM_SIZE) &&
+        (program_info->header.version <= EBPF_PROGRAM_INFORMATION_VERSION_LATEST) &&
```
Comment (dthaler) [inline libs/shared/shared_common.c:52]: BUG (see above)

File: libs/shared/shared_common.c [hunk 11]
```diff
@@ -1,8 +1,72 @@
...
+        ebpf_validate_helper_function_prototype_array(
+            program_info->global_helper_prototype, program_info->count_of_global_helpers));
+}
+
+bool
+ebpf_validate_program_section_info(_In_ const ebpf_program_section_info_t* section_info)
+{
+    return (
+        (section_info != NULL) && (section_info->header.size >= sizeof(ebpf_program_section_info_t)) &&
+        (section_info->header.version <= EBPF_PROGRAM_SECTION_VERSION_LATEST) && (section_info->section_name != NULL) &&
```
Comment (dthaler) [inline libs/shared/shared_common.c:66]: BUG (see above)

**Review by dthaler — CHANGES_REQUESTED** [review]
Please fix logic bugs before merging

File: docs/eBpfExtensions.md (hunk 3 above)
Comment (shankarseal) [reply docs/eBpfExtensions.md]: I modified the text here. Version numbers are updated for breaking changes. Older types will no longer be supported.

File: libs/execution_context/ebpf_native.c (hunk 6 above)
Comment (shankarseal) [reply libs/execution_context/ebpf_native.c:124]: Because [MSDN](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/ns-netioddk-_npi_provider_characteristics) says:
`The version of the NMR with which the provider is registering. A provider module should set this member to zero.`

File: libs/shared/shared_common.c (hunk 8 above)
Comment (shankarseal) [reply libs/shared/shared_common.c:34]: Context descriptor is used in `EbpfHelperPrototype`. If you think this should be versioned, I will create a separate issue to track this.

File: libs/shared/shared_common.c (hunk 9 above)
Comment (shankarseal) [reply libs/shared/shared_common.c]: Fixed here and elsewhere.

File: libs/execution_context/ebpf_program.c [hunk 12]
```diff
@@ -182,61 +183,153 @@ _IRQL_requires_max_(PASSIVE_LEVEL) static ebpf_result_t _ebpf_program_compute_pr
...
+    if (program_data == NULL) {
+        EBPF_LOG_MESSAGE(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_PROGRAM,
+            "The program information provider data is not present.");
+        goto Done;
+    }
+
+    // Verify that that the program information version is supported.
+    if (program_data->header.version > EBPF_PROGRAM_DATA_VERSION_LATEST ||
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:240]: Suggest
```suggestion
    if (program_data->header.version != EBPF_PROGRAM_DATA_VERSION_LATEST ||
```

Since if VERSION_LATEST ever changes to 1 then this line will need to be updated, so it's best to cause a failure if that happens.

**Review by dthaler — DISMISSED** [review]
Approved but with a suggestion

**Review by dthaler — APPROVED** [review]

File: libs/execution_context/ebpf_program.c [hunk 13]
```diff
@@ -182,61 +183,153 @@ _IRQL_requires_max_(PASSIVE_LEVEL) static ebpf_result_t _ebpf_program_compute_pr
...
+    if (program_data == NULL) {
+        EBPF_LOG_MESSAGE(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_PROGRAM,
+            "The program information provider data is not present.");
+        goto Done;
+    }
+
+    // Verify that that the program information version is supported.
+    if (!(program_data->header.version == EBPF_PROGRAM_DATA_VERSION_LATEST) &&
```
Comment (matthewige) [inline libs/execution_context/ebpf_program.c:240]: What is the expected flow if we have a newer version extension than what the eBPF runtime supports? The extension would fail to register, and then it's just out of luck?

Do we need to provide a way to query the version at runtime? Then the extension can 'downgrade' itself before registering.

File: libs/execution_context/ebpf_program.c [hunk 14]
```diff
@@ -182,61 +183,153 @@ _IRQL_requires_max_(PASSIVE_LEVEL) static ebpf_result_t _ebpf_program_compute_pr
...
+        (program_data->header.size >= EBPF_PROGRAM_DATA_VERSION_0_MINIMUM_SIZE)) {
```
Comment (matthewige) [inline libs/execution_context/ebpf_program.c:241]: Do we want to make _LATEST #defines for these size definitions too?

File: docs/eBpfExtensions.md [hunk 15]
```diff
@@ -189,7 +199,17 @@ This optional function is used to populate the flat buffers representing the dat
 returned to the application when the `bpf_prog_test_run_opts` call completes. In addition, the function frees any
 resources allocated in the `ebpf_program_context_create_t` call.
 
-### 2.2 Program Information NPI Client Attach and Detach Callbacks
+### 2.2 Backward compatibility of the Extension data structures
+All the extension data structures are versioned. New fields can be added to the end of a data structure to maintain backward compatibility
+with the existing extensions. In such cases, the size field of the header will be updated but the version of the structure *will not change*.
+Existing eBPF extensions will continue to work without requiring recompilation.
```
Comment (saxena-anurag) [inline docs/eBpfExtensions.md:205]: In such a case, I think we should add explicit padding in all these structs to ensure compiler does not automatically add padding and increase the size? (and also enabled compiler warning `C4820`)?

File: include/ebpf_program_types.h [hunk 16]
```diff
@@ -4,48 +4,51 @@
...
+// when version is EBPF_PROGRAM_TYPE_DESCRIPTOR_VERSION_0.
 typedef struct _ebpf_program_type_descriptor
 {
+    ebpf_extension_header_t header;
     const char* name;
     const ebpf_context_descriptor_t* context_descriptor;
     GUID program_type;
     uint32_t bpf_prog_type;
     char is_privileged;
 } ebpf_program_type_descriptor_t;
```
Comment (saxena-anurag) [inline include/ebpf_program_types.h:22]: nit: should we instead define this struct as `ebpf_program_type_descriptor_t_0` or `ebpf_program_type_descriptor_t_version_0` and then:

```
#define ebpf_program_type_descriptor_t ebpf_program_type_descriptor_t_version_0
```
Whenever there is a version bump in future, we will need to define a `version_1` struct and also update the above `#define`?

File: libs/execution_context/ebpf_program.c (hunk 13 above)
Comment (dthaler) [reply libs/execution_context/ebpf_program.c:240]: The version number should almost never change (ideally never), only the size field should change as the structure is extended, so it's not a big deal either way.  That said, if it ever does then I expect ebpf_program.c to accept both the old and the new versions, at least until some future time when support for the old version would be removed completely.

File: libs/store_helper/ebpf_store_helper.c [hunk 17]
```diff
@@ -144,6 +149,11 @@ ebpf_store_update_section_information(
     for (uint32_t i = 0; i < section_info_count; i++) {
         ebpf_store_key_t section_key = NULL;
 
+        if (!ebpf_validate_program_section_info(&section_info[i])) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/store_helper/ebpf_store_helper.c:152]: I don't think it means that.  Multiple versions can be accepted as valid at the same time.

File: docs/eBpfExtensions.md (hunk 15 above)
Comment (dthaler) [reply docs/eBpfExtensions.md:205]: I would recommend against explicit padding at the end since it would unduly bloat the structure.  For example, say you have 1 byte of non-header data in a struct at present, and later want to add a second 1 byte field.  If it's padded already then you cannot reuse that space for the new field since you have to add more space to get the size to increase.  If on the other hand the struct is packed (no padding) then you can easily add just 1 more byte and all is well.

File: include/ebpf_program_types.h (hunk 16 above)
Comment (dthaler) [reply include/ebpf_program_types.h:22]: That sounds like a fine suggestion to me.

File: docs/eBpfExtensions.md (hunk 15 above)
Comment (saxena-anurag) [reply docs/eBpfExtensions.md:205]: The proposal in this doc is that if new fields are to be added, the version need not be incremented, and just the size of the struct should be sufficient to determine whether the struct is of older or newer "version". 

This logic will break if compiler had earlier added padding and adding a new field _does not_ increase the struct size (as the new field got absorbed in the padding). In such a case, just the struct size may not be sufficient to determine if there is an extra field in the struct.

Comment (dthaler) [reply docs/eBpfExtensions.md:205]: Use the [`pack` pragma](https://learn.microsoft.com/en-us/cpp/preprocessor/pack?view=msvc-170) to guarantee that it won't be padded.

Comment (saxena-anurag) [reply docs/eBpfExtensions.md:205]: One minor concern with using `pack` pragma is that the fields may not be aligned, and can possibly cause performance issues, but I think in our case NMR interface objects will be in hot path.

File: netebpfext/net_ebpf_ext_bind.c [hunk 18]
```diff
@@ -57,28 +57,30 @@ _ebpf_bind_context_destroy(
 // Bind Program Information NPI Provider.
 //
 static ebpf_program_data_t _ebpf_bind_program_data = {
+    .header = {EBPF_PROGRAM_DATA_VERSION_0, sizeof(ebpf_program_data_t)},
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_bind.c:60]: No. The `LATEST` macro is not even exposed to extensions. Extensions should use the version corresponding to the `ebpf_windows.h` that they are including.

File: include/ebpf_program_types.h (hunk 16 above)
Comment (shankarseal) [reply include/ebpf_program_types.h:22]: There were two options:
1) Change the type identifier when version changed and support older versions.
2) Version numbers need not be changed, even if data structure changes, as long as they are backward compatible. Version changes will be made only when there is a breaking change, and there will be a corresponding major version change in eBPF as well, at which point older extensions/clients will no longer work.

We chose the second option (which is the style used for NDIS structs) - where the type identifier does not change even when struct definition changes maintaining backward compatibility. A version change indicates breaking change and extensions/clients need to be re-compiled.

File: docs/eBpfExtensions.md (hunk 15 above)
Comment (shankarseal) [reply docs/eBpfExtensions.md:205]: @dthaler - that is why I think having a minor version in addition the length field will be useful. And we could maintain an expected length per version. Such as:

`#define EBPF_HELPER_FUNCTION_PROTOTYPE_VERSION_1_2_EXPECTED_LENGTH offsetof(ebpf_helper_function_prototype_t, the_new_field`) + sizeof(typeof(the_new_field))`.

If you are OK with this, I can publish another iteration of the changes. If not, I will pack to 1 byte which may have performance issues.

File: libs/store_helper/ebpf_store_helper.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/store_helper/ebpf_store_helper.c:152]: No @dthaler - not quite. eBPF v1.N can update an extension struct and maintain backward compatibility with the structs shipped in versions v1.0 - v1.N-1 as by definition such changes will be backward compatible. However, once there is a change in the struct that is no longer backward compatible (such as adding a field in the middle or changing the type of a field), then we must change the version of the struct and will be considered a breaking change. We will then have to bump the version of the product to v2.0.

File: libs/execution_context/ebpf_program.c (hunk 14 above)
Comment (shankarseal) [reply libs/execution_context/ebpf_program.c:241]: Agreed. I will make the changes in the next iteration. Keeping the comment active.

File: libs/execution_context/ebpf_program.c (hunk 13 above)
Comment (shankarseal) [reply libs/execution_context/ebpf_program.c:240]: See above comments. A version change is associated with a breaking change, which by definition would not be backward compatible.

File: docs/eBpfExtensions.md (hunk 15 above)
Comment (dthaler) [reply docs/eBpfExtensions.md:205]: I suspect the minor version concept is more bug prone, and we didn't go that route for NMR.  Just using sizeof(the_struct) seems less bug prone to me.  I'd rather either pack to 1 byte, or use explicit padding as Anurag suggested.  (And you can use explicit internal padding to ensure alignment if you do go the pack route.)  But I think Anurag's suggestion would be preferable to having a minor version.   That's my 2 cents.

File: libs/store_helper/ebpf_store_helper.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/store_helper/ebpf_store_helper.c:152]: Once you bump the version of a structure, it doesn't mean that code cannot accept two versions of the same structure.  It can.  Thus a newer version of ebpfapi.dll can support an older extension.   My hope though is that we never need to bump the version.

File: docs/eBpfExtensions.md (hunk 15 above)
Comment (shankarseal) [reply docs/eBpfExtensions.md]: I chose to move ahead with the minor version and no packing.

File: include/ebpf_windows.h [hunk 19]
```diff
@@ -49,3 +60,50 @@ typedef enum _ebpf_helper_function
...
+ * @brief Header of an eBPF extension data structure.
+ * Every eBPF extension data structure must start with this header.
+ * New fields can be added to the end of an eBPF extension data structure
+ * without breaking backward compatibility. The minor version field must be incremented in such cases. The major version
+ * field must be updated only if the new data structure is not backward compatible.
+ */
+typedef struct _ebpf_extension_header
+{
+    uint16_t major_version; ///< Major version of the extension data structure.
+    uint16_t minor_version; ///< Minor version of the extension data structure.
```
Comment (dthaler) [inline include/ebpf_windows.h:107]: I still see no value in having a minor_version field in this structure.  It's completely redundant with the `size` field in my opinion, and adding it here increases the opportunity for bugs.  In a sense the `size` field IS a minor version already so we don't need a second one.

Comment (shankarseal) [reply include/ebpf_windows.h:107]: The `size` field will be ambiguous due to compiler padding structs. Either we have to add explicit "reserved" fields to pad or pack at 1 byte and incur perf hit.

File: libs/shared/shared_common.c [hunk 20]
```diff
@@ -1,8 +1,170 @@
...
+
+    if (header->major_version != _supported_ebpf_extension_major_version[object_type]) {
+        goto Done;
+    }
+
+    if (header->minor_version > _latest_ebpf_extension_minor_version[object_type]) {
+        goto Done;
+    }
+
+    if (header->size != ebpf_get_extension_data_size(object_type, header->major_version, header->minor_version)) {
```
Comment (matthewige) [inline libs/shared/shared_common.c:89]: Should we also have a check that header->size != 0?

File: docs/eBpfExtensions.md [hunk 21]
```diff
@@ -62,8 +62,16 @@ initialized as follows:
 * `NpiSpecificCharacteristics`: Pointer to structure of type `ebpf_extension_data_t`.
   * The `data` field of this structure should point to a structure of type `ebpf_program_data_t`.
 
+#### `ebpf_extension_header_t` Struct
+This is a mandatory header that is common to all data structures needed by eBPF extensions to register with the eBPF framework.
+* `major_version`: Major version of the extension data structure.
+* `minor_version`: Minor version of the extension data structure.
+* `size`: Size of the extension data structure.
+ When populating these data structures, extensions must set the correct major and minor versions of the data structure. The set of supported version numbers for the various extension structures are listed in `ebpf_windows.h`. The size field should be set to `sizeof()` of the corresponding type.
```
Comment (matthewige) [inline docs/eBpfExtensions.md:70]: should the last sentence instead read 'The size field MUST be set to...'

File: include/ebpf_windows.h (hunk 19 above)
Comment (dthaler) [reply include/ebpf_windows.h:107]: Not quite.  I'm saying size should NOT be set to sizeof(struct), it should be set to FIELD_END(last_member).  If you do this, there's no reason to have a minor version, and it's less bug prone.

Comment (shankarseal) [reply include/ebpf_windows.h:107]: If the last field in the struct is a bitfield, and in the next iteration another bitfield is added (such as #3344) ; what macro can be used to detect the change in size?

Comment (dthaler) [reply include/ebpf_windows.h:107]: For the case where the bits are always zero until later a use for them is added where the zero value is backwards compatible, then bit fields can be put into a named struct, and then you just use the same FIELD_END macro on the named struct.

If a zero value is not backwards compatible, I will argue that adding such fields into existing bytes is a bug-prone way to do additions over time, so I'd disallow it.  In that case the fact that no macro can be used is a good thing.

Bottom line is that 3344 ought to put bitfields into a named struct that sizeof() can work with.

Comment (shankarseal) [reply include/ebpf_windows.h]: Accepted your proposal. Please check.

File: libs/api_common/store_helper_internal.cpp [hunk 22]
```diff
@@ -15,77 +16,225 @@ ebpf_store_key_t root_registry_key_current_user = HKEY_CURRENT_USER;
...
+
+    EBPF_LOG_ENTRY();
+
+    ebpf_result_t result = ebpf_read_registry_value_dword(data_key, EBPF_EXTENSION_HEADER_VERSION, &version);
+    if (result != EBPF_SUCCESS) {
+        return result;
+    }
+    result = ebpf_read_registry_value_dword(data_key, EBPF_EXTENSION_HEADER_SIZE, &size);
+    if (result != EBPF_SUCCESS) {
+        return result;
```
Comment (dthaler) [inline libs/api_common/store_helper_internal.cpp:44]: It looks to me like lines 40 and 44 won't log the exit even though line 36 logged the entry.
Line 50 will but only on success.

File: docs/eBpfExtensions.md [hunk 23]
```diff
@@ -373,15 +401,15 @@ To operate on the eBPF store, the user mode application needs to link with eBPFA
         _In_reads_(section_info_count) const ebpf_program_section_info_t* section_info, uint32_t section_info_count);
     ```
 
-- `ebpf_store_update_program_information`: updates program information in the eBPF store, given a pointer to an array of program information (i.e., `_ebpf_program_info`):
+- `ebpf_store_update_program_data`: updates program information in the eBPF store, given a pointer to an array of program information (i.e., `_ebpf_program_info`):
 
     ```c
     ebpf_result_t
-    ebpf_store_update_program_information(
+    ebpf_store_update_program_data(
```
Comment (saxena-anurag) [inline docs/eBpfExtensions.md:408]: nit: since this API takes a pointer to `ebpf_program_info_t`, should the name of API also be aligned with similar name, i.e. `ebpf_store_update_program_information` instead of `ebpf_store_update_program_data`?

File: include/ebpf_windows.h [hunk 24]
```diff
@@ -49,3 +61,44 @@ typedef enum _ebpf_helper_function
     EBPF_UPDATE_ELEMENT = 2, ///< Update map element.
     EBPF_DELETE_ELEMENT = 3, ///< Delete a map element.
 } ebpf_helper_function_t;
+
+#define EBPF_MAX_GENERAL_HELPER_FUNCTION 0xFFFF
+
+// Version 1 of the eBPF extension data structures and their lengths.
+#define EBPF_ATTACH_CLIENT_DATA_CURRENT_VERSION 1
```
Comment (saxena-anurag) [inline include/ebpf_windows.h:68]: question: what is the guidance when there is a breaking change? will all these versions be bumped to `2` or only versions of those structs which had a breaking change will be bumped to 2 and remaining will remain at `1`

File: libs/shared/shared_common.c [hunk 25]
```diff
@@ -1,8 +1,151 @@
...
+
+uint16_t _supported_ebpf_extension_version[] = {
+    EBPF_PROGRAM_TYPE_DESCRIPTOR_CURRENT_VERSION,
+    EBPF_HELPER_FUNCTION_PROTOTYPE_CURRENT_VERSION,
+    EBPF_PROGRAM_INFORMATION_CURRENT_VERSION,
+    EBPF_PROGRAM_DATA_CURRENT_VERSION,
+    EBPF_PROGRAM_SECTION_INFORMATION_CURRENT_VERSION,
+};
+
+#define EBPF_PROGRAM_TYPE_DESCRIPTOR_SIZE_0 EBPF_OFFSET_OF(ebpf_program_type_descriptor_t, is_privileged) + sizeof(char)
```
Comment (saxena-anurag) [inline libs/shared/shared_common.c:26]: nit: Should we define and use a common macro (something like below) for these?
```
#define EBPF_FIELD_OFFSET(s, m) (EBPF_OFFSET_OF(s, m) + sizeof(((s*)0)->m))
```

**Review by dthaler — APPROVED** [review]

File: libs/shared/shared_common.c [hunk 26]
```diff
@@ -1,8 +1,151 @@
...
+size_t _ebpf_program_type_descriptor_supported_size[] = {EBPF_PROGRAM_TYPE_DESCRIPTOR_SIZE_0};
+
+#define EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0 \
+    EBPF_OFFSET_OF(ebpf_helper_function_prototype_t, arguments) + 5 * sizeof(ebpf_argument_type_t)
+size_t _ebpf_helper_function_prototype_supported_size[] = {EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0};
```
Comment (saxena-anurag) [inline libs/shared/shared_common.c:31]: It seems we are defining the macro for structure sizes twice. For example,

```
#define EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0 \
    EBPF_OFFSET_OF(ebpf_helper_function_prototype_t, arguments) + 5 * sizeof(ebpf_argument_type_t)
```
and
```
#define EBPF_HELPER_FUNCTION_PROTOTYPE_CURRENT_VERSION_SIZE \
    (EBPF_OFFSET_OF(ebpf_helper_function_prototype_t, arguments) + 5 * sizeof(ebpf_argument_type_t))
```

I think we can define them only once, and `EBPF_HELPER_FUNCTION_PROTOTYPE_CURRENT_VERSION_SIZE` can be defined to `EBPF_HELPER_FUNCTION_PROTOTYPE_SIZE_0`

File: include/ebpf_windows.h (hunk 24 above)
Comment (dthaler) [reply include/ebpf_windows.h:68]: Only those with breaking changes.

File: libs/shared/shared_common.c (hunk 25 above)
Comment (dthaler) [reply libs/shared/shared_common.c:26]: a common macro is a good idea

**Comment by saxena-anurag** [comment]
Should we try to include this PR in 0.15 release? That will help validate (via regression tests) that the versioning changes in this PR are backward compatible when #3344 is included in 0.16.

File: libs/shared/shared_common.c (hunk 26 above)
Comment (shankarseal) [reply libs/shared/shared_common.c:31]: This is by design, because with non-breaking changes, we will update the second macro to have _SIZE_N+1 with a different offset.

File: libs/api_common/store_helper_internal.cpp [hunk 27]
```diff
@@ -15,77 +16,225 @@ ebpf_store_key_t root_registry_key_current_user = HKEY_CURRENT_USER;
...
     HKEY helper_store_key,
     _In_z_ const wchar_t* helper_name,
     _Out_ ebpf_helper_function_prototype_t* helper_prototype) noexcept
 {
     int32_t status;
     ebpf_result_t result = EBPF_SUCCESS;
     HKEY helper_info_key = nullptr;
+    uint32_t offset = 0;
+    char serialized_data[sizeof(ebpf_helper_function_prototype_t)] = {0};
+    size_t expected_size = sizeof(helper_prototype->helper_id) + sizeof(helper_prototype->return_type) +
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/api_common/store_helper_internal.cpp:64]: Call to `ebpf_validate_program_info` in line 378 should enforce that they are the same.

**Comment by shankarseal** [comment]
> Should we try to include this PR in 0.15 release? That will help validate (via regression tests) that the versioning changes in this PR are backward compatible when #3344 is included in 0.16.

No.  let us please wait till 0.16.

File: docs/eBpfExtensions.md (hunk 23 above)
Comment (shankarseal) [reply docs/eBpfExtensions.md]: renamed.

**Review by dthaler — DISMISSED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]
